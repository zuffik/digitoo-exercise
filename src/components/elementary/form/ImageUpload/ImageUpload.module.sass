@import "../../../../styles/variables"

.delete
  color: $danger
  padding: 0
  margin: 0
  border: none
  outline: none !important
  cursor: pointer
  background-color: transparent
