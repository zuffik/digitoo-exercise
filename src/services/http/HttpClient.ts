import axios, {AxiosError, AxiosInstance} from 'axios';
import {storage} from "../wrappers/LocalStorage";

export const createHttpClient = (): AxiosInstance => {
    const auth = storage.getItem('auth', {});
    const client = axios.create({
        baseURL: process.env.REACT_APP_API_URL,
        responseType: 'json',
        headers: {
            'X-API-KEY': process.env.REACT_APP_API_KEY,
            ...auth.accessToken && {'Authorization': `${auth.accessToken.token_type} ${auth.accessToken.access_token}`}
        },
    });
    client.interceptors.response.use(undefined, (e: AxiosError) => {
        if (e.response?.status === 401) {
            // just logout the user unless you want to keep credentials in storage
            storage.setItem('auth', {...auth, accessToken: undefined});
        }
        throw e;
    });
    return client;
};

export const httpClient: AxiosInstance = createHttpClient();
