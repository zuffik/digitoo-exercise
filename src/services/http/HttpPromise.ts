import {AxiosPromise, AxiosResponse} from "axios";
import {HttpState} from "../types/HttpState";
import {update} from "../wrappers/Update";

export const handleHttpPromise = <T>(promise: AxiosPromise<T>, setter: (state: (state: HttpState<T>) => HttpState<T>) => void, successCallback?: (response: AxiosResponse<T>) => void) => {
    setter(article => update(article, {loading: {$set: true}}))
    promise
        .then(response => {
            setter(article => update(article, {$set: {loading: false, data: response.data}}))
            successCallback?.(response);
        })
        .catch(error => setter(article => update(article, {$set: {loading: false, error}})));
}
