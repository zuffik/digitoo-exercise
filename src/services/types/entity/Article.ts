export interface Article {
  articleId: string;
  title: string;
  perex: string;
  imageId: string;
  createdAt: string;
  lastUpdatedAt: string;
}
