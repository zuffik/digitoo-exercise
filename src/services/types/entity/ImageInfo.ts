export interface ImageInfo {
    imageId: string;
    name: string;
}
