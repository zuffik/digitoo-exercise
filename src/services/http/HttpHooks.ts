import {makeUseAxios, Options} from 'axios-hooks';
import {httpClient} from './HttpClient';
import {AxiosRequestConfig} from 'axios';

export const useAxios = makeUseAxios({cache: undefined, axios: httpClient});
export const useGet = <T>(url: string, config?: Omit<AxiosRequestConfig, 'method' | 'url'>, opts?: Options) =>
    useAxios<T>({
        ...config,
        method: 'get',
        url,
    });
export const usePost = <T>(
    url: string,
    config?: Omit<AxiosRequestConfig, 'method' | 'url'>,
    opts?: Options
) =>
    useAxios<T>({
        ...config,
        method: 'post',
        url,
    });
