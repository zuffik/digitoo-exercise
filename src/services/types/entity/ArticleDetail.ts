import { Article } from "./Article";
import { Comment } from "./Comment";

export interface ArticleDetail extends Article {
  content: string;
  comments: Comment[];
}
