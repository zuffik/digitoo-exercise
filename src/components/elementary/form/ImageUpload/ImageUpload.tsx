import * as React from 'react';
import {FileImage} from "../../FileImage/FileImage";
import styles from './ImageUpload.module.sass';
import {useIsStorybook} from "../../../storybook/StorybookContext";
import {ImageInfo} from "../../../../services/types/entity/ImageInfo";
import {Spinner} from "../../progress/Spinner/Spinner";
import {defaultHttpState, HttpState} from "../../../../services/types/HttpState";
import {handleHttpPromise} from "../../../../services/http/HttpPromise";
import {httpClient} from "../../../../services/http/HttpClient";

interface Props {
    imageId?: string;
    onUploadFinished?: (image: ImageInfo) => void;
}

export const ImageUpload: React.FC<Props> = ({onUploadFinished, ...props}: Props) => {
    const [file, setFile] = React.useState<undefined | File>();
    const [uploading, setUploading] = React.useState<HttpState<ImageInfo[]>>(defaultHttpState);
    const isStorybook = useIsStorybook();

    const onRemoveImage = () => {
        // 405: not allowed on freshly created image
        // DELETE `/images/${imageId}`
        setFile(undefined);
    }
    const onFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const f = e.target.files?.[0];
        if (!isStorybook) {
            // upload new image
            const fd = new FormData();
            fd.append('image', f as Blob);
            handleHttpPromise(httpClient.post(`/images`, fd), setUploading, response => {
                onUploadFinished?.(response.data[0]);
            });
        }
        setFile(f);
        e.target.value = '';
    };
    return (
        <>
            {
                !file
                    ? <label htmlFor="upload" data-testid="image-upload-large" className="btn btn-secondary">Upload an Image</label>
                    : (
                        <>
                            <FileImage data-testid="image-upload-img" file={file}/>
                            <div className="d-flex flex-row align-items-center mt-3">
                                <label htmlFor="upload" data-testid="image-upload-small" className="text-primary m-0">Upload new</label>
                                <span className="text-muted mx-2"> | </span>
                                <button className={styles.delete} onClick={onRemoveImage} data-testid="image-upload-remove">Delete</button>
                            </div>
                        </>
                    )
            }
            <input type="file" className="d-none" id="upload" data-testid="image-upload-input" onChange={onFileChange}/>
            {uploading.loading && <Spinner data-testid="image-upload-spinner" size={16}/>}
        </>
    );
};
