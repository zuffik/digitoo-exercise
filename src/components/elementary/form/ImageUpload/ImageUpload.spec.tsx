import * as React from 'react';
import {fireEvent, render} from '@testing-library/react'
import {ImageUpload} from "./ImageUpload";
import {httpClient} from "../../../../services/http/HttpClient";
import {ImageInfo} from "../../../../services/types/entity/ImageInfo";

describe('<ImageUpload/>', () => {
    it('should upload image', () => {
        const onUploadFinished = jest.fn();
        URL.createObjectURL = jest.fn(() =>'source')
        const spyUpload = jest.spyOn(httpClient, 'post').mockImplementation(jest.fn(async () => [{
            name: 'name',
            imageId: '123'
        } as ImageInfo]))
        const {asFragment, getByTestId} = render(<ImageUpload onUploadFinished={onUploadFinished}/>);
        expect(getByTestId('image-upload-large')).toBeInTheDOM();
        const file = new File([], 'image.png', {type: 'image/png'});
        fireEvent.change(getByTestId('image-upload-input'), {
            target: {
                files: [file]
            }
        });
        expect(spyUpload).toHaveBeenCalledWith('/images', expect.any(FormData));
        expect(getByTestId('image-upload-small')).toBeInTheDOM();
        expect(getByTestId('image-upload-img')).toBeInTheDOM();
        expect(asFragment()).toMatchSnapshot();
    });
    it('should upload image and remove it', () => {
        const onUploadFinished = jest.fn();
        URL.createObjectURL = jest.fn(() =>'source')
        const spyUpload = jest.spyOn(httpClient, 'post').mockImplementation(jest.fn(async () => [{
            name: 'name',
            imageId: '123'
        } as ImageInfo]))
        const {asFragment, getByTestId, queryByTestId} = render(<ImageUpload onUploadFinished={onUploadFinished}/>);
        const file = new File([], 'image.png', {type: 'image/png'});
        fireEvent.change(getByTestId('image-upload-input'), {
            target: {
                files: [file]
            }
        });
        expect(spyUpload).toHaveBeenCalledWith('/images', expect.any(FormData));
        fireEvent.click(getByTestId('image-upload-remove'));
        expect(getByTestId('image-upload-large')).toBeInTheDOM();
        expect(queryByTestId('image-upload-remove')).toBeNull();
        expect(queryByTestId('image-upload-small')).toBeNull();
        expect(queryByTestId('image-upload-img')).toBeNull();
        expect(asFragment()).toMatchSnapshot();
    });
});
