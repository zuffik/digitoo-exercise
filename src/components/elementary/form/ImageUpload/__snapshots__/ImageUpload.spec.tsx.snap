// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageUpload/> should upload image 1`] = `
<DocumentFragment>
  <img
    data-testid="image-upload-img"
    src="source"
  />
  <div
    class="d-flex flex-row align-items-center mt-3"
  >
    <label
      class="text-primary m-0"
      data-testid="image-upload-small"
      for="upload"
    >
      Upload new
    </label>
    <span
      class="text-muted mx-2"
    >
       | 
    </span>
    <button
      class="delete"
      data-testid="image-upload-remove"
    >
      Delete
    </button>
  </div>
  <input
    class="d-none"
    data-testid="image-upload-input"
    id="upload"
    type="file"
  />
  <div
    class="spinner-border"
    size="16"
    style="width: 16px; height: 16px;"
  />
</DocumentFragment>
`;

exports[`<ImageUpload/> should upload image and remove it 1`] = `
<DocumentFragment>
  <label
    class="btn btn-secondary"
    data-testid="image-upload-large"
    for="upload"
  >
    Upload an Image
  </label>
  <input
    class="d-none"
    data-testid="image-upload-input"
    id="upload"
    type="file"
  />
  <div
    class="spinner-border"
    size="16"
    style="width: 16px; height: 16px;"
  />
</DocumentFragment>
`;
